File: backend/app/models/transport.py
"""Transport models: Bus, Route, BusStop, GPSLog."""
from sqlalchemy import (
    ARRAY,
    Boolean,
    Index,
    Column,
    Date,
    DateTime,
    Enum,
    Float,
    ForeignKey,
    Integer,
    Numeric,
    String,
    Text,
    Time,
    func,
    text,
)
from sqlalchemy.dialects.postgresql import JSONB, UUID
from sqlalchemy.orm import relationship

from app.models.base import SchoolModel


class Route(SchoolModel):
    __tablename__ = "routes"

    name = Column(String(200), nullable=False)
    description = Column(Text)
    distance_km = Column(Numeric(6, 2))
    estimated_time_mins = Column(Integer)
    is_active = Column(Boolean, default=True)


class Bus(SchoolModel):
    __tablename__ = "buses"

    vehicle_number = Column(String(20), nullable=False)
    driver_id = Column(UUID(as_uuid=True), ForeignKey("users.id"))
    conductor_id = Column(UUID(as_uuid=True), ForeignKey("users.id"))
    capacity = Column(Integer)
    current_students_count = Column(Integer, default=0)
    gps_device_id = Column(String(100))
    make = Column(String(100))
    model = Column(String(100))
    year = Column(Integer)
    insurance_expiry = Column(DateTime)
    route_id = Column(UUID(as_uuid=True), ForeignKey("routes.id"))
    is_active = Column(Boolean, default=True)

    driver = relationship("User", foreign_keys=[driver_id])
    conductor = relationship("User", foreign_keys=[conductor_id])
    route = relationship("Route", backref="buses")


class BusStop(SchoolModel):
    __tablename__ = "bus_stops"

    route_id = Column(UUID(as_uuid=True), ForeignKey("routes.id"), nullable=False)
    name = Column(String(200), nullable=False)
    name_nepali = Column(String(200))
    latitude = Column(Numeric(10, 7))
    longitude = Column(Numeric(10, 7))
    sequence_number = Column(Integer)
    arrival_time_am = Column(Time)
    arrival_time_pm = Column(Time)
    student_ids = Column(ARRAY(UUID(as_uuid=True)))

    route = relationship("Route", backref="stops")


class GPSLog(SchoolModel):
    __tablename__ = "gps_logs"

    bus_id = Column(UUID(as_uuid=True), ForeignKey("buses.id"), nullable=False)
    latitude = Column(Numeric(10, 7), nullable=False)
    longitude = Column(Numeric(10, 7), nullable=False)
    speed_kmh = Column(Float)
    heading = Column(Float)
    accuracy_m = Column(Float)
    timestamp = Column(DateTime, nullable=False)
    firebase_synced = Column(Boolean, default=False)

    bus = relationship("Bus", backref="gps_logs")


# ══════════════════════════════════════════════════════════════════════════
# S-A4 (A-10/A-11) — the 4-layer trip lifecycle (the SchoolBusTrack model,
# rebuilt for multi-tenant Postgres):
#   definition (TransportTrip) → daily instance (TransportTripInstance)
#   → per-stop planned/actual (TransportTripInstanceStop)
#   → per-student ride (TransportTripReservation, ride_status 0/1/2/3/4)
# plus per-student notification prefs (A-11) and the alert dedupe ledger.
# Timestamps are TIMESTAMPTZ (their TIME-only columns lose the date).
# ══════════════════════════════════════════════════════════════════════════

class TransportTrip(SchoolModel):
    """A recurring trip definition: one route run morning or afternoon on
    the selected weekdays, starting on effective_date_bs. Suspend = cancel
    instances for specific dates (holidays) via instances' status."""

    __tablename__ = "transport_trips"

    route_id = Column(UUID(as_uuid=True), ForeignKey("routes.id"), nullable=False, index=True)
    bus_id = Column(UUID(as_uuid=True), ForeignKey("buses.id"))
    driver_id = Column(UUID(as_uuid=True), ForeignKey("users.id"))
    direction = Column(
        Enum("morning", "afternoon", name="trip_direction"), nullable=False
    )
    effective_date_bs = Column(String(10))
    # ISO weekday ints present in the schedule, e.g. [0..6] (Mon=0) — []
    # = every day.
    weekdays = Column(JSONB, default=list)
    first_stop_time = Column(Time)          # local planned start
    stop_to_stop_avg_mins = Column(Integer, default=5)
    status = Column(
        Enum("active", "retired", name="trip_status"), default="active", index=True
    )
    name = Column(String(200))

    route = relationship("Route")
    bus = relationship("Bus")
    driver = relationship("User")


class TransportTripInstance(SchoolModel):
    """One concrete run on one date (materialized by the publish beat).
    Carries driver/bus snapshots so history survives reassignment, the
    last GPS fix (what parents render instantly), and a per-instance socket
    room name (private, JWT-scoped — never the public channels SBT used)."""

    __tablename__ = "transport_trip_instances"
    __table_args__ = (
        Index(
            "uq_transport_trip_instances_trip_date",
            "trip_id", "date",
            unique=True,
            postgresql_where=text("is_deleted = false"),
        ),
    )

    trip_id = Column(UUID(as_uuid=True), ForeignKey("transport_trips.id"), nullable=False, index=True)
    date = Column(Date, nullable=False)
    date_bs = Column(String(10))
    direction = Column(Enum("morning", "afternoon", name="trip_direction"), nullable=False)
    driver_id = Column(UUID(as_uuid=True), ForeignKey("users.id"))
    bus_id = Column(UUID(as_uuid=True), ForeignKey("buses.id"))
    status = Column(
        Enum("scheduled", "running", "completed", "cancelled", name="trip_instance_status"),
        default="scheduled", index=True,
    )
    started_at = Column(DateTime(timezone=True))
    ended_at = Column(DateTime(timezone=True))
    last_lat = Column(Numeric(10, 7))
    last_lng = Column(Numeric(10, 7))
    last_speed_kmh = Column(Float)
    last_fix_at = Column(DateTime(timezone=True))

    trip = relationship("TransportTrip", backref="instances")
    bus = relationship("Bus")
    driver = relationship("User")
    stops = relationship(
        "TransportTripInstanceStop", back_populates="instance",
        cascade="all, delete-orphan", order_by="TransportTripInstanceStop.seq",
    )


class TransportTripInstanceStop(SchoolModel):
    """Per-stop planned/actual pair. planned_ts derives from the trip's
    first_stop_time + per-stop offsets; actual_ts is stamped by the geofence
    engine (arrival) — TIMESTAMPTZ, never TIME-only."""

    __tablename__ = "transport_trip_instance_stops"

    instance_id = Column(
        UUID(as_uuid=True), ForeignKey("transport_trip_instances.id"),
        nullable=False, index=True,
    )
    stop_id = Column(UUID(as_uuid=True), ForeignKey("bus_stops.id"), nullable=False)
    seq = Column(Integer, nullable=False)
    planned_ts = Column(DateTime(timezone=True))
    actual_ts = Column(DateTime(timezone=True))

    instance = relationship("TransportTripInstance", back_populates="stops")
    stop = relationship("BusStop")


class TransportTripReservation(SchoolModel):
    """The board/alight register — ride_status IS the transport attendance:
    0 waiting, 1 onboard, 2 missed, 3 dropped, 4 absent (pre-marked)."""

    __tablename__ = "transport_trip_reservations"
    __table_args__ = (
        Index(
            "uq_transport_reservations_instance_student",
            "instance_id", "student_id",
            unique=True,
            postgresql_where=text("is_deleted = false"),
        ),
    )

    instance_id = Column(
        UUID(as_uuid=True), ForeignKey("transport_trip_instances.id"),
        nullable=False, index=True,
    )
    student_id = Column(UUID(as_uuid=True), ForeignKey("students.id"), nullable=False, index=True)
    start_stop_id = Column(UUID(as_uuid=True), ForeignKey("bus_stops.id"))
    end_stop_id = Column(UUID(as_uuid=True), ForeignKey("bus_stops.id"))
    ride_status = Column(Integer, nullable=False, default=0, index=True)
    boarded_at = Column(DateTime(timezone=True))
    dropped_at = Column(DateTime(timezone=True))
    fee_status = Column(String(20))     # ok|overdue|none — gate on the fees plugin

    instance = relationship("TransportTripInstance", backref="reservations")
    student = relationship("Student")


class TransportNotificationPref(SchoolModel):
    """Per-student notification toggles + approach radius (A-11). One row
    per student; missing row = sensible defaults (everything on, 150 m)."""

    __tablename__ = "transport_notification_prefs"

    student_id = Column(UUID(as_uuid=True), ForeignKey("students.id"), nullable=False, index=True)
    near_pickup_radius_m = Column(Integer, default=150)
    near_dropoff_radius_m = Column(Integer, default=150)
    # 7 event toggles (bus_arrived_at_school is a dead switch in SBT — we
    # model only events the engine actually fires):
    notify_next_stop_pickup = Column(Boolean, default=True)
    notify_near_pickup = Column(Boolean, default=True)
    notify_arrived_pickup = Column(Boolean, default=True)
    notify_picked_up = Column(Boolean, default=True)
    notify_missed_pickup = Column(Boolean, default=True)
    notify_near_dropoff = Column(Boolean, default=True)
    notify_arrived_dropoff = Column(Boolean, default=True)

    student = relationship("Student")


class TransportAlertLog(SchoolModel):
    """Dedupe ledger: one alert per (instance, student, event) inside the
    dedupe window — a driver circling a stop must not spam guardians."""

    __tablename__ = "transport_alert_log"

    instance_id = Column(UUID(as_uuid=True), ForeignKey("transport_trip_instances.id"), index=True)
    student_id = Column(UUID(as_uuid=True), ForeignKey("students.id"), index=True)
    event = Column(String(40), nullable=False)
    sent_at = Column(DateTime(timezone=True), nullable=False, default=func.now())

File: aschool_shared/lib/models/transport.dart
/// Transport Model
import '../utils/safe_parse.dart';

class TransportRoute {
  final String id;
  final String title;
  final String? vehicleId;
  final String? vehicleNumber;
  final String? driverName;
  final String? driverPhone;
  final double? fareAmount;
  final List<TransportStop> stops;

  const TransportRoute({
    required this.id,
    required this.title,
    this.vehicleId,
    this.vehicleNumber,
    this.driverName,
    this.driverPhone,
    this.fareAmount,
    this.stops = const [],
  });

  factory TransportRoute.fromJson(Map<String, dynamic> json) {
    return TransportRoute(
      id: safeString(json['id']),
      title: safeString(json['title']),
      vehicleId: safeStringOrNull(json['vehicle_id']),
      vehicleNumber: safeStringOrNull(json['vehicle_number']),
      driverName: safeStringOrNull(json['driver_name']),
      driverPhone: safeStringOrNull(json['driver_phone']),
      fareAmount: safeDoubleOrNull(json['fare_amount']),
      stops: safeMapList(json['stops'] ?? json['route_stops'])
          .map(TransportStop.fromJson)
          .toList(),
    );
  }
}

class TransportStop {
  final String id;
  final String name;
  final String? pickTime;
  final String? dropTime;
  final double? distance;
  final double? additionalFare;
  final double? latitude;
  final double? longitude;

  const TransportStop({
    required this.id,
    required this.name,
    this.pickTime,
    this.dropTime,
    this.distance,
    this.additionalFare,
    this.latitude,
    this.longitude,
  });

  factory TransportStop.fromJson(Map<String, dynamic> json) {
    return TransportStop(
      id: safeString(json['id']),
      name: safeString(json['name'], fallback: safeString(json['stop_name'])),
      pickTime: safeStringOrNull(json['pick_time']),
      dropTime: safeStringOrNull(json['drop_time']),
      distance: safeDoubleOrNull(json['distance']),
      additionalFare: safeDoubleOrNull(json['additional_fare']),
      latitude: safeDoubleOrNull(json['latitude']),
      longitude: safeDoubleOrNull(json['longitude']),
    );
  }
}

// ── S-A4 trip lifecycle ────────────────────────────────────────────────────
// Daily TripInstances + per-stop planned/actual + ride register, as served
// by /transport/instances and /transport/instances/<id> (see backend
// app/api/v1/transport.py).

/// One day's run of a trip definition: morning/afternoon, scheduled →
/// running → completed (or cancelled), with per-stop and per-student state.
class TripInstance {
  final String id;
  final String tripId;
  final String date;
  final String? dateBs;
  final String direction; // morning | afternoon
  final String? driverId;
  final String? busId;
  final String? bus;
  final String status; // scheduled | running | completed | cancelled
  final DateTime? startedAt;
  final DateTime? endedAt;
  final double? lastLat;
  final double? lastLng;
  final double? lastSpeedKmh;
  final DateTime? lastFixAt;
  final List<InstanceStop> stops;
  final List<TripPassenger> passengers;

  const TripInstance({
    required this.id,
    required this.tripId,
    required this.date,
    required this.direction,
    required this.status,
    this.dateBs,
    this.driverId,
    this.busId,
    this.bus,
    this.startedAt,
    this.endedAt,
    this.lastLat,
    this.lastLng,
    this.lastSpeedKmh,
    this.lastFixAt,
    this.stops = const [],
    this.passengers = const [],
  });

  bool get isRunning => status == 'running';
  bool get isCompleted => status == 'completed';
  bool get isScheduled => status == 'scheduled';

  int get visitedStopCount => stops.where((s) => s.visited).length;
  int get totalStopCount => stops.length;
  int get onboardCount =>
      passengers.where((p) => p.rideStatus == TripPassenger.statusOnboard).length;

  factory TripInstance.fromJson(Map<String, dynamic> json) {
    return TripInstance(
      id: safeString(json['id']),
      tripId: safeString(json['trip_id']),
      date: safeString(json['date']),
      dateBs: safeStringOrNull(json['date_bs']),
      direction: safeString(json['direction'], fallback: 'morning'),
      driverId: safeStringOrNull(json['driver_id']),
      busId: safeStringOrNull(json['bus_id']),
      bus: safeStringOrNull(json['bus']),
      status: safeString(json['status'], fallback: 'scheduled'),
      startedAt: safeDateTime(json['started_at']),
      endedAt: safeDateTime(json['ended_at']),
      lastLat: safeDoubleOrNull(json['last_lat']),
      lastLng: safeDoubleOrNull(json['last_lng']),
      lastSpeedKmh: safeDoubleOrNull(json['last_speed_kmh']),
      lastFixAt: safeDateTime(json['last_fix_at']),
      stops: safeMapList(json['stops']).map(InstanceStop.fromJson).toList()
        ..sort((a, b) => a.seq.compareTo(b.seq)),
      passengers:
          safeMapList(json['passengers']).map(TripPassenger.fromJson).toList(),
    );
  }
}

/// A stop as planned/executed on a specific instance (planned vs actual
/// arrival). Visited = the run has stamped an actual timestamp.
class InstanceStop {
  final String id;
  final String stopId;
  final String? stopName;
  final int seq;
  final double? lat;
  final double? lng;
  final DateTime? plannedTs;
  final DateTime? actualTs;

  const InstanceStop({
    required this.id,
    required this.stopId,
    required this.seq,
    this.stopName,
    this.lat,
    this.lng,
    this.plannedTs,
    this.actualTs,
  });

  bool get visited => actualTs != null;

  factory InstanceStop.fromJson(Map<String, dynamic> json) {
    return InstanceStop(
      id: safeString(json['id']),
      stopId: safeString(json['stop_id']),
      stopName: safeStringOrNull(json['stop_name']),
      seq: safeInt(json['seq']),
      lat: safeDoubleOrNull(json['lat']),
      lng: safeDoubleOrNull(json['lng']),
      plannedTs: safeDateTime(json['planned_ts']),
      actualTs: safeDateTime(json['actual_ts']),
    );
  }
}

/// One student's ride on an instance (ride register row).
class TripPassenger {
  static const int statusWaiting = 0;
  static const int statusOnboard = 1;
  static const int statusMissed = 2;
  static const int statusDropped = 3;

  final String id;
  final String studentId;
  final String? studentName;
  final String? startStopId;
  final String? endStopId;
  final int rideStatus;
  final DateTime? boardedAt;
  final DateTime? droppedAt;

  const TripPassenger({
    required this.id,
    required this.studentId,
    required this.rideStatus,
    this.studentName,
    this.startStopId,
    this.endStopId,
    this.boardedAt,
    this.droppedAt,
  });

  String get rideStatusLabel {
    switch (rideStatus) {
      case statusOnboard:
        return 'Onboard';
      case statusMissed:
        return 'Missed';
      case statusDropped:
        return 'Dropped';
      default:
        return 'Waiting';
    }
  }

  factory TripPassenger.fromJson(Map<String, dynamic> json) {
    return TripPassenger(
      id: safeString(json['id']),
      studentId: safeString(json['student_id']),
      studentName: safeStringOrNull(json['student_name']),
      startStopId: safeStringOrNull(json['start_stop_id']),
      endStopId: safeStringOrNull(json['end_stop_id']),
      rideStatus: safeInt(json['ride_status']),
      boardedAt: safeDateTime(json['boarded_at']),
      droppedAt: safeDateTime(json['dropped_at']),
    );
  }
}

/// Per-student transport notification toggles + geofence radii
/// (/transport/notification-prefs). Missing rows on the backend mean the
/// defaults, so the client mirrors that: a pref row may be synthesized.
class TransportNotificationPref {
  final String studentId;
  final int nearPickupRadiusM;
  final int nearDropoffRadiusM;
  final bool notifyNextStopPickup;
  final bool notifyNearPickup;
  final bool notifyArrivedPickup;
  final bool notifyPickedUp;
  final bool notifyMissedPickup;
  final bool notifyNearDropoff;
  final bool notifyArrivedDropoff;

  const TransportNotificationPref({
    required this.studentId,
    this.nearPickupRadiusM = 150,
    this.nearDropoffRadiusM = 150,
    this.notifyNextStopPickup = true,
    this.notifyNearPickup = true,
    this.notifyArrivedPickup = true,
    this.notifyPickedUp = true,
    this.notifyMissedPickup = true,
    this.notifyNearDropoff = true,
    this.notifyArrivedDropoff = true,
  });

  TransportNotificationPref copyWith({
    String? studentId,
    int? nearPickupRadiusM,
    int? nearDropoffRadiusM,
    bool? notifyNextStopPickup,
    bool? notifyNearPickup,
    bool? notifyArrivedPickup,
    bool? notifyPickedUp,
    bool? notifyMissedPickup,
    bool? notifyNearDropoff,
    bool? notifyArrivedDropoff,
  }) {
    return TransportNotificationPref(
      studentId: studentId ?? this.studentId,
      nearPickupRadiusM: nearPickupRadiusM ?? this.nearPickupRadiusM,
      nearDropoffRadiusM: nearDropoffRadiusM ?? this.nearDropoffRadiusM,
      notifyNextStopPickup:
          notifyNextStopPickup ?? this.notifyNextStopPickup,
      notifyNearPickup: notifyNearPickup ?? this.notifyNearPickup,
      notifyArrivedPickup: notifyArrivedPickup ?? this.notifyArrivedPickup,
      notifyPickedUp: notifyPickedUp ?? this.notifyPickedUp,
      notifyMissedPickup: notifyMissedPickup ?? this.notifyMissedPickup,
      notifyNearDropoff: notifyNearDropoff ?? this.notifyNearDropoff,
      notifyArrivedDropoff:
          notifyArrivedDropoff ?? this.notifyArrivedDropoff,
    );
  }

  /// PUT /transport/notification-prefs body for this student.
  Map<String, dynamic> toRequestJson() => {
        'student_id': studentId,
        'near_pickup_radius_m': nearPickupRadiusM,
        'near_dropoff_radius_m': nearDropoffRadiusM,
        'notify_next_stop_pickup': notifyNextStopPickup,
        'notify_near_pickup': notifyNearPickup,
        'notify_arrived_pickup': notifyArrivedPickup,
        'notify_picked_up': notifyPickedUp,
        'notify_missed_pickup': notifyMissedPickup,
        'notify_near_dropoff': notifyNearDropoff,
        'notify_arrived_dropoff': notifyArrivedDropoff,
      };

  factory TransportNotificationPref.fromJson(
    Map<String, dynamic> json, {
    int defaultPickupRadiusM = 150,
    int defaultDropoffRadiusM = 150,
  }) {
    return TransportNotificationPref(
      studentId: safeString(json['student_id']),
      nearPickupRadiusM:
          safeInt(json['near_pickup_radius_m'], fallback: defaultPickupRadiusM),
      nearDropoffRadiusM: safeInt(json['near_dropoff_radius_m'],
          fallback: defaultDropoffRadiusM),
      notifyNextStopPickup:
          safeBool(json['notify_next_stop_pickup'], fallback: true),
      notifyNearPickup: safeBool(json['notify_near_pickup'], fallback: true),
      notifyArrivedPickup:
          safeBool(json['notify_arrived_pickup'], fallback: true),
      notifyPickedUp: safeBool(json['notify_picked_up'], fallback: true),
      notifyMissedPickup:
          safeBool(json['notify_missed_pickup'], fallback: true),
      notifyNearDropoff: safeBool(json['notify_near_dropoff'], fallback: true),
      notifyArrivedDropoff:
          safeBool(json['notify_arrived_dropoff'], fallback: true),
    );
  }
}

/// Response bundle of GET /transport/notification-prefs.
class TransportPrefsBundle {
  final List<TransportNotificationPref> prefs;
  final int defaultPickupRadiusM;
  final int defaultDropoffRadiusM;

  const TransportPrefsBundle({
    required this.prefs,
    this.defaultPickupRadiusM = 150,
    this.defaultDropoffRadiusM = 150,
  });

  factory TransportPrefsBundle.fromJson(Map<String, dynamic> json) {
    final defaults = safeMap(json['defaults']);
    return TransportPrefsBundle(
      prefs:
          safeMapList(json['prefs']).map(TransportNotificationPref.fromJson).toList(),
      defaultPickupRadiusM: safeInt(
          defaults['near_pickup_radius_m'], fallback: 150),
      defaultDropoffRadiusM: safeInt(
          defaults['near_dropoff_radius_m'], fallback: 150),
    );
  }
}

/// Response bundle of GET /transport/instances?date_bs=….
class TransportInstancesPage {
  final String date;
  final List<TripInstance> instances;

  const TransportInstancesPage({required this.date, required this.instances});

  factory TransportInstancesPage.fromJson(Map<String, dynamic> json) {
    return TransportInstancesPage(
      date: safeString(json['date']),
      instances: safeMapList(json['instances'])
          .map(TripInstance.fromJson)
          .toList()
        ..sort((a, b) {
          // Running first, then scheduled, then finished/cancelled.
          int rank(String s) => switch (s) {
                'running' => 0,
                'scheduled' => 1,
                'completed' => 2,
                _ => 3,
              };
          final byStatus = rank(a.status).compareTo(rank(b.status));
          if (byStatus != 0) return byStatus;
          return a.direction.compareTo(b.direction);
        }),
    );
  }
}
